File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/ContextModifiers/SignatureModifierNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;

namespace SoundPlayground.Parser.AbstractSyntaxTree.ContextModifiers {
    public class SignatureModifierNode : ContextModifierNode {
        public int? Upper { get; set; } = null;

        public int? Lower { get; set; } = null;

        public SignatureModifierNode () {}

        public SignatureModifierNode ( int? upper = null, int? lower = null ) {
            Upper = upper;
            Lower = lower;
        }

        public SignatureModifierNode ( ( int, int ) signature ) : this( signature.Item1, signature.Item2 ) {
            
        }

        public override void Modify ( Context context ) {
            if ( Upper != null && Lower != null ) {
                context.TimeSignature = ( Upper.Value, Lower.Value );
            } else if ( Upper != null ) {
                context.TimeSignature = ( Upper.Value, context.TimeSignature.Item2 );
            } else if ( Lower != null ) {
                context.TimeSignature = ( context.TimeSignature.Item1, Lower.Value );
            }
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/context_modifiers/signature_modifier_node.py
from .context_modifier_node import ContextModifierNode
from core.events import ContextChangeEvent
from core import Context, Voice

class SignatureModifierNode( ContextModifierNode ):
    def __init__ ( self, upper = None, lower = None, position : (int, int) = None ):
        super().__init__( position )

        self.upper = upper
        self.lower = lower

    def apply ( self, voice : Voice ):
        if self.upper != None or self.lower != None:
            voice.time_signature = ( self.upper, self.lower )
        elif self.upper != None:
            context.time_signature = ( self.upper, context.time_signature[ 1 ] )
        elif self.lower != None:
            context.time_signature = ( context.time_signature[ 0 ], self.lower )

    def modify ( self, context : Context ):
        if self.upper != None or self.lower != None:
            context.voice = context.voice.clone( time_signature = ( self.upper, self.lower ) )
        elif self.upper != None:
            context.voice = context.voice.clone( time_signature = ( self.upper, context.time_signature[ 1 ] ) )
        elif self.lower != None:
            context.voice = context.voice.clone( time_signature = ( context.time_signature[ 0 ], self.lower ) )

        yield ContextChangeEvent( context.cursor, "timeSignature", context.voice.time_signature )
